geo/src/algorithm/contains/polygon.rs

```diff
@@ -0,0 +1,91 @@
+use super::Contains;
+use crate::intersects::Intersects;
+use crate::kernels::*;
+use crate::*;
+
+// ┌─────────────────────────────┐
+// │ Implementations for Polygon │
+// └─────────────────────────────┘
+
+impl<T> Contains<Coordinate<T>> for Polygon<T>
+where
+    T: HasKernel,
+{
+    fn contains(&self, coord: &Coordinate<T>) -> bool {
+        match utils::coord_pos_relative_to_ring(*coord, &self.exterior()) {
+            utils::CoordPos::OnBoundary | utils::CoordPos::Outside => false,
+            _ => self.interiors().iter().all(|ls| {
+                utils::coord_pos_relative_to_ring(*coord, ls) == utils::CoordPos::Outside
+            }),
+        }
+    }
+}
+
+impl<T> Contains<Point<T>> for Polygon<T>
+where
+    T: HasKernel,
+{
+    fn contains(&self, p: &Point<T>) -> bool {
+        self.contains(&p.0)
+    }
+}
+
+// TODO: ensure DE-9IM compliance: esp., when
+// line.start and line.end is on the boundaries
```

Comment:
Right as your TODO alludes, it seems like the current logic would fail for a case like:

![image](https://user-images.githubusercontent.com/217057/93694015-da81f600-fabb-11ea-9da2-8234cee21441.png)

Where the line touches, but does not pass, the interior.
Or if the line touches, but does not pass, the exterior.

Reply:
Yeah.  Also see the test `line_in_polygon_edgecases` in `contains/mod.rs`.  We might be able to use the winding order trait and a few additions to contains to solve these cases, but wanted to attempt this in a different PR.